### PR TITLE
[DK][APB-967] secure PUT and DELETE endpoints.

### PR DESCRIPTION
**Pull request by dimitraka71**

### CODE REVIEW

File: app/uk/gov/hmrc/agentclientrelationships/controllers/actions/AuthActions.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import uk.gov.hmrc.play.http.HeaderCarrier
+import play.api.mvc._
+
+import scala.concurrent.Future
+
+trait AuthActions {
+  me: Results =>
+  def authConnector: AuthConnector
+
+  protected val withAuthority = new ActionBuilder[RequestWithAuthority] with ActionRefiner[Request, RequestWithAuthority] {
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/actions/AuthActions.scala]: withAuthority can be private

File: app/uk/gov/hmrc/agentclientrelationships/controllers/actions/AuthActions.scala [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    protected def refine[A](request: RequestWithAuthority[A]): Future[Either[Result, AgentOrClient[A]]] = {
+      implicit val hc = HeaderCarrier.fromHeadersAndSession(request.headers, None)
+      (for {
+        opArn <- authConnector.currentArn(request.authority.enrolmentsUrl)
+        opMtdItId <- authConnector.currentMtdItId(request.authority.enrolmentsUrl)
+      } yield (opArn, opMtdItId) match {
+        case (None, None) => Left(NoAgentOrClient)
+        case (optArn, optMtdItId) => Right(AgentOrClient(optArn, optMtdItId, request))
+      }).recover({
+        case e: uk.gov.hmrc.play.http.Upstream4xxResponse if e.upstreamResponseCode == 401 =>
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/actions/AuthActions.scala]: import Upstream4xxResponse?

**Comment by dimitraka71** [comment]
fixed

File: app/uk/gov/hmrc/agentclientrelationships/controllers/ErrorResults.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.agentclientrelationships.controllers
+
+import play.api.libs.json.Json.toJson
+import play.api.libs.json.{JsValue, Json, Writes}
+import play.api.mvc.Results.{Forbidden, Unauthorized}
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/ErrorResults.scala]: organise imports

File: app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala [hunk 4]
```diff
@@ -74,7 +77,7 @@ class Relationships @Inject()(service: RelationshipsService) extends BaseControl
     }
   }
 
-  def create(arn: Arn, mtdItId: MtdItId): Action[AnyContent] = Action.async { implicit request =>
+  def create(arn: Arn, mtdItId: MtdItId): Action[AnyContent] = agentOrClient.async { implicit request =>
```
Comment (PaulHodgson) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala]: We need to check that the arn or mtditid supplied to the method match the arn or mtditid obtained from the session.

So if the user is an agent the 2 arms must be the same, else if a client the 2 client ids must match.

See agent-client-authorisation AgencyInvitationsController.forThisAgency() for an example arn check.

**Comment by frankoid** [comment]
Commit comments don't reference JIRA - please fix this before merging (I can help with the requisite git history editing if required).

File: app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala [hunk 5]
```diff
@@ -74,42 +78,54 @@ class Relationships @Inject()(service: RelationshipsService) extends BaseControl
...
-    } yield ())
-      .map(_ => NoContent)
-      .recover {
-        case ex =>
-          Logger.warn(s"Could not delete relationship: ${ex.getMessage}")
-          NotFound(toJson(ex.getMessage))
-      }
+  private def forThisAgentOrClient(arn: Arn, mtdItId: MtdItId)
+                                  (block: => Future[Result])(implicit request: AgentOrClient[_]) = {
+    (arn != request.arn.getOrElse(""), mtdItId != request.mtdItId.getOrElse("")) match {
```
Comment (cavster) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala]: I personally don't like using .getOrElse   but I guess it doesn't matter ¯\_(ツ)_/¯

Comment (cavster) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala]: LGTM

Comment (frankoid) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala]: I'm not keen on this particular use of `getOrElse` either, will push a refactoring

Comment (frankoid) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala]: @cavster I've refactored forThisAgentOrClient and also added a unit test for it

Comment (frankoid) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala]: @cavster please can you re-review, if the whole PR looks good to you please merge

**Comment by frankoid** [comment]
N.B. I have force pushed so if you have a local branch please delete it

File: app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala [hunk 6]
```diff
@@ -74,42 +79,55 @@ class Relationships @Inject()(service: RelationshipsService) extends BaseControl
...
-      _ <- service.checkForRelationship(mtdItId, agentCode)
-      _ <- service.deleteRelationship(arn, mtdItId)
-    } yield ())
-      .map(_ => NoContent)
-      .recover {
-        case ex =>
-          Logger.warn(s"Could not delete relationship: ${ex.getMessage}")
-          NotFound(toJson(ex.getMessage))
-      }
+  private[controllers] def forThisAgentOrClient(requiredArn: Arn, requiredMtdItId: MtdItId)
```
Comment (cavster) [inline app/uk/gov/hmrc/agentclientrelationships/controllers/Relationships.scala]: LGTM

**Pull request merged**